src/App.tsx

```diff
@@ -8,27 +8,89 @@ import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
 
+import { getTodos } from './api';
+
+import { Todo } from './types/Todo';
+
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [visibleTodos, setVisibleTodos] = useState<Todo[]>(todos);
+  const [isLoading, setIsLoading] = useState(true);
+  const [openedTodo, setOpenedTodo] = useState<null | Todo>(null);
+  const [query, setQuery] = useState('');
+  const [todoCategory, setTodosCategory] = useState('all');
+
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos()
+      .then((fetchedTodos) => {
+        setTodos(fetchedTodos);
+        setVisibleTodos(fetchedTodos);
+      })
+      .catch((error) => error.message)
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  useEffect(() => {
+    let filteredTodos = todos;
+
+    switch (todoCategory) {
+      case 'completed':
+        filteredTodos = todos.filter(todo => todo.completed);
+        break;
+      case 'active':
+        filteredTodos = todos.filter(todo => !todo.completed);
+        break;
+      default:
+    }
+
+    if (query) {
+      const queryToLowerCase = query.toLowerCase();
+
+      filteredTodos = filteredTodos.filter(todo => todo.title.toLowerCase().includes(queryToLowerCase));
+    }
+
+    setVisibleTodos(filteredTodos);
+  }, [query, todoCategory, todos]);
+
+  const toggleTodoModal = (todo: null | Todo) => {
```

Comment:
one more thing - name functions based on requirements from checklist pt 3
```suggestion
  const handleTodoModalToggle = (todo: null | Todo) => {
```